src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Utilities/MetadataExtensions.cs

```diff
@@ -0,0 +1,80 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Utilities
+{
+    internal static class MetadataExtensions
+    {
+        /// <summary>
+        /// Finds the resolved reference item for a given unresolved reference.
+        /// </summary>
+        /// <param name="projectRuleSnapshot">Resolved reference project items snapshot to search.</param>
+        /// <param name="itemSpec">The unresolved reference item name.</param>
+        /// <returns>The key is item name and the value is the metadata dictionary.</returns>
+        public static IImmutableDictionary<string, string> GetProjectItemProperties(this IProjectRuleSnapshot projectRuleSnapshot, string itemSpec)
+        {
+            Requires.NotNull(projectRuleSnapshot, nameof(projectRuleSnapshot));
+            Requires.NotNullOrEmpty(itemSpec, nameof(itemSpec));
+
+            projectRuleSnapshot.Items.TryGetValue(itemSpec, out IImmutableDictionary<string, string> properties);
+
+            return properties;
```

Comment:
Do we want to explicitly return null if `TryGetValue` fails?

Reply:
If `TryGetValue` fails, `properties` is definitely `null`. Are you saying the code could be clearer?

Reply:
Yes, `properties` being bull is an implementation detail.  I assume we want to be explicit that we intend to return null

---
In reply to: [231938165](https://github.com/dotnet/project-system/pull/4242#discussion_r231938165) [](ancestors = 231938165)

Reply:
I'll make it explicit.